packages/core-app-api/src/lib/AuthConnector/PinnipedAuthConnector.ts

```diff
@@ -0,0 +1,239 @@
+
+let warned = false;
+
+type Options<AuthSession> = {
+  /**
+   * DiscoveryApi instance used to locate the auth backend endpoint.
+   */
+  discoveryApi: DiscoveryApi;
+  /**
+   * Environment hint passed on to auth backend, for example 'production' or 'development'
+   */
+  environment: string;
+  /**
+   * Information about the auth provider to be shown to the user.
+   * The ID Must match the backend auth plugin configuration, for example 'google'.
+   */
+  provider: AuthProviderInfo;
+  /**
+   * Function used to transform an auth response into the session type.
+   */
+  sessionTransform?(response: any): AuthSession | Promise<AuthSession>;
+  /**
+   * ConfigApi instance used to configure authentication flow of pop-up or redirect.
+   */
+  configApi?: ConfigApi;
+  /**
+   * Options used to configure auth popup
+   */
+  popupOptions?: PopupOptions;
+  audience: string;
+};
+
+/**
+ * PinnipedAuthConnector is the auth connector to consume Pinniped. It talks to the
+ * backend auth plugin through the standardized API, and requests user permission
+ * via the OAuthRequestApi.
+ */
+export class PinnipedAuthConnector<AuthSession>
+  implements AuthConnector<AuthSession>
+{
+  private readonly discoveryApi: DiscoveryApi;
+  private readonly environment: string;
+  private readonly audience: string;
+  private readonly provider: AuthProviderInfo;
+  private readonly sessionTransform: (response: any) => Promise<AuthSession>;
+  private readonly enableExperimentalRedirectFlow: boolean;
+  private readonly popupOptions: PopupOptions | undefined;
+  constructor(options: Options<AuthSession>) {
+    const {
+      configApi,
+      discoveryApi,
+      environment,
+      provider,
+      audience,
+      sessionTransform = id => id,
+      popupOptions,
+    } = options;
+
+    if (!warned && !configApi) {
+      // eslint-disable-next-line no-console
+      console.warn(
+        'DEPRECATION WARNING: Authentication providers require a configApi instance to configure the authentication flow. Please provide one to the authentication provider constructor.',
+      );
+      warned = true;
+    }
+
+    this.enableExperimentalRedirectFlow = configApi
+      ? configApi.getOptionalBoolean('enableExperimentalRedirectFlow') ?? false
+      : false;
+
+    this.discoveryApi = discoveryApi;
+    this.environment = environment;
+    this.provider = provider;
+    this.sessionTransform = sessionTransform;
+    this.popupOptions = popupOptions;
+    this.audience = audience;
+  }
+
+  async createSession(options: CreateSessionOptions): Promise<AuthSession> {
+    if (options.instantPopup) {
+      if (this.enableExperimentalRedirectFlow) {
+        return this.executeRedirect();
+      }
+      return this.showPopup();
+    }
+    return this.showPopup();
+  }
+
+  async refreshSession(): Promise<any> {
+    const res = await fetch(
+      await this.buildUrl('/refresh', {
+        optional: true,
+      }),
```

Comment:
to properly get cluster-scoped ID tokens, we're going to need to pass the audience here
```suggestion
      await this.buildUrl('/refresh', {
        optional: true,
        audience: this.audience,
      }),
```

Reply:
Done !